### PR TITLE
Issue263 : Added model filter in neighbordb

### PR DESCRIPTION
**Pull request by az-blip**
Hello,
Like some other people asked for, we needed to filter definitions by switch model. In our case, it was for EOS version compatibility with older switches.
We tried to add the feature in this small patch. This allow to add a "model" parameter in the neighbordb patterns, as a regular expression.
We're actually running our forked version in production, and everything seems fine.
We also added information about this filter in the "documentation" section.

Christophe and Alain
University of Strasbourg (FR)

### CODE REVIEW

File: ztpserver/controller.py [hunk 1]
```diff
@@ -218,7 +218,7 @@ def put_config(self, request, **kwargs):
 
         fobj = None
         filename = self.expand(node_id, STARTUP_CONFIG_FN)
-        body = str(request.body)
+        body = request.body.decode("utf-8")
```
Comment (dlobato) [inline ztpserver/controller.py:221]: Is this change required?

Comment (az-blip) [reply ztpserver/controller.py:221]: Hi,
It's not related to the model filter, but it's a bugfix to the backup of the switches configurations. This bug was probably introduced with the python3 migration due to the default usage of unicode.
Without this fix, the startup-config is saved as a one-line file when you post the startup-config from the switch to the ztpserver. The saved file wasn't re-usable to restore the swich configuration.

Comment (dlobato) [reply ztpserver/controller.py:221]: I've created a separate PR #407 to fix this

**Review by dlobato — COMMENTED** [review]
Have a look at the tests I'd added to check if I'm checking the right use cases

File: docs/config.rst [hunk 2]
```diff
@@ -476,7 +476,8 @@ unique_id of the node) and global patterns.
 Rules:
 
  - if multiple node-specific entries reference the same unique_id, only the first will be in effect - all others will be ignored
- - if both the **node** and **interfaces** attributes are specified and a node's unique_id is a match, but the topology information is not, then the overall match will fail and the global patterns will not be considered
+ - if the **node** and **interfaces** or **model** attributes are specified and a node's unique_id is a match, but the topology information or model information are not, then the overall match will fail and the global patterns will not be considered
+ - **model** can't be the only matching attributes, **node** or **interfaces** are mandatory to have a valid match
```
[the remark this answers is not in the document]
Comment (az-blip) [reply docs/config.rst:480]: Hi,
Sorry for the delay, I was in vacation.

> If I'm understanding this correctly, shouldn't node and model be mutually exclusive? I don't see the case where both are needed. So I'd rephrase this as:

Yes, I think it's a nonsense to use the two together, but it won't result in an error so we thought it was best to explain what would happen.

> * if both the **node** and **interfaces** attributes are specified and a node's unique_id is a match, but the topology information is not, then the overall match will fail and the global patterns will not be considered
> * if both the **model** and **interfaces** attributes are specified and a node's model is a match, but the topology information is not, then the overall match will fail and the global patterns will not be considered
> * **node** and **model** can't be specified in the same pattern

OK for me.

> The model can't be the only matching attr is not fulfilled with the current implementation. What's the reasoning for this rule?

As I remember, it's was mainly because we didn't really needed this use case and this allowed to only operate a very minor modification in the code.

**Review by dlobato — APPROVED** [review]
